File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/templates/TemplateScrollableContent.kt
package com.perrystreet.woof.designsystem.atomic.templates

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawing
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListScope
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.SpacingRoles
import com.perrystreet.woof.designsystem.theme.Theme

@Composable
fun TemplateScrollableContent(
    topBar: @Composable () -> Unit,
    content: LazyListScope.() -> Unit,
) {
    Scaffold(
        containerColor = Theme.colors.background,
        contentWindowInsets = WindowInsets.safeDrawing,
        topBar = {
            Box(modifier = Modifier.padding(horizontal = PaddingRoles.Screen.Regular.dp)) {
                topBar()
            }
        },
    ) { paddingValues ->
        LazyColumn(
            modifier =
                Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
            contentPadding =
                PaddingValues(
                    start = PaddingRoles.Screen.Regular.dp,
                    end = PaddingRoles.Screen.Regular.dp,
                    top = PaddingRoles.Screen.Compact.dp,
                    bottom = PaddingRoles.Screen.Regular.dp,
                ),
            verticalArrangement = Arrangement.spacedBy(SpacingRoles.Module.Compact.dp),
            content = content,
        )
    }
}

File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Templates/TemplateScrollableContent.swift
import SwiftUI

public struct TemplateScrollableContent<TopBar: View, Content: View>: View {
    @Environment(\.theme) private var theme

    private let topBar: () -> TopBar
    private let content: () -> Content

    public init(@ViewBuilder topBar: @escaping () -> TopBar, @ViewBuilder content: @escaping () -> Content) {
        self.topBar = topBar
        self.content = content
    }

    public var body: some View {
        VStack(spacing: 0) {
            topBar()
                .padding(.horizontal, PaddingRoles.Screen.regular.rawValue)
            ScrollView {
                LazyVStack(alignment: .leading, spacing: SpacingRoles.Module.compact.rawValue) {
                    content()
                }
                .padding(.horizontal, PaddingRoles.Screen.regular.rawValue)
                .padding(.top, PaddingRoles.Screen.compact.rawValue)
                .padding(.bottom, PaddingRoles.Screen.regular.rawValue)
            }
            .scrollIndicators(.hidden)
        }
        .background(theme.colors.background.ignoresSafeArea())
    }
}
